### PR TITLE
Fix missing slashes in URLs and allow switching emulator account

### PR DESCRIPTION
**Pull request by arnodb**
Hi,

this PR fixes potentially missing slashes in URLs (I had that problem with emulator URLs), and allows to specify the emulator account in an added constructor.

Note that I have not tested the queue and filesystem URLs with Azurite yet, but I suspect they do not work. This PR does not regress from that point of view.

Cheers!

### CODE REVIEW

**Comment by arnodb** [comment]
It appears that the queue URL is incorrect for Azurite. Moreover, the more arguments are added to the constructor, the more it is not safe. I would like your views on how to arrange that, and will be happy to implement them.

File: sdk/storage/src/queue/requests/clear_messages_builder.rs [hunk 1]
```diff
@@ -28,7 +28,10 @@ impl<'a> ClearMessagesBuilder<'a> {
     pub async fn execute(
         &self,
     ) -> Result<ClearMessagesResponse, Box<dyn std::error::Error + Sync + Send>> {
-        let mut url = self.queue_client.queue_url()?.join("messages")?;
+        let mut url = self.queue_client.queue_url()?.to_owned();
+        url.path_segments_mut()
+            .map_err(|_| "Invalid queue URL")?
+            .push("messages");
```
Comment (rylev) [inline sdk/storage/src/queue/requests/clear_messages_builder.rs:34]: This code is being repeated a lot. Is there anyway to share this in all the places it's being used?

**Review by rylev — CHANGES_REQUESTED** [review]
Thanks! 😊 I think this is a good move, but I wonder if we can reduce the change to repeating less code.

File: sdk/storage/src/queue/requests/clear_messages_builder.rs (hunk 1 above)
Comment (arnodb) [reply sdk/storage/src/queue/requests/clear_messages_builder.rs:34]: Maybe, but if you have a close look, you'll see that sometimes I did not want to change the returned error type and used `.map_err(|_| url::ParseError::SetHostOnCannotBeABaseUrl)`. This is not fully correct, but it is the closest variant I found to raise an error which should never happen as the URL was parsed already.

If you want to have a bit of harmony, what would be your preffered implementation?

Comment (arnodb) [reply sdk/storage/src/queue/requests/clear_messages_builder.rs]: Meanwhile, I pushed the beginning of the refactoring, tell me what you think of it.

**Comment by arnodb** [comment]
Hi, what can I do to make this pull request progress?

**Comment by ctaggart** [comment]
Hey @arnodb, thank for the pull request. I'll be reviewing this week and will try to get this unstuck.

**Review by ctaggart — APPROVED** [review]
I'll tackle the errors in another pull request. The refactor looks good. I created #277 to fix the conflicts. If you merge them into this pull request, I'll merge it.

**Comment by arnodb** [comment]
I merged #277 in this PR as requested.

Before you merge, may I draw your attention on the 9 `map_err` calls which could probably be harmonized if you tell me which version you prefer?

**Comment by arnodb** [comment]
I merged upstream to fix conflicts.

**Comment by ctaggart** [comment]
@arnodb, thanks for fixing the conflicts. It looks like the tests are passing on master. I'll see if I can reproduce the failed tests:

test table::clients::entity_client::integration_tests::test_merge ... FAILED
test table::clients::entity_client::integration_tests::test_update ... FAILED
test table::clients::partition_key_client::integration_tests::test_transaction ... FAILED
test table::clients::table_client::integration_tests::test_create_delete ... FAILED
test table::clients::table_client::integration_tests::test_insert ... FAILED

**Comment by ctaggart** [comment]
I was able to reproduce this locally. Those tests are only failing on this branch, not master. It is pretty easy to reproduce on Windows:
```
&"C:\Program Files (x86)\Microsoft SDKs\Azure\Storage Emulator\AzureStorageEmulator.exe" start
cargo test --features test_integration
```

**Comment by arnodb** [comment]
I don't have a windows emulator and Azurite seems not to support a few things required by tests. So I'm going to try a few commits.

**Comment by arnodb** [comment]
Green :heavy_check_mark:.

**Review by ctaggart — APPROVED** [review]

**Pull request merged**